daemon/docker_watcher.go

```diff
@@ -123,19 +124,33 @@ func (d *Daemon) processEvent(m dTypesEvents.Message) {
 }
 
 func getCiliumEndpointID(cont *dTypes.ContainerJSON, gwIP *addressing.NodeAddress) uint16 {
-	for _, contNetwork := range cont.NetworkSettings.Networks {
-		ipv6gw := net.ParseIP(contNetwork.IPv6Gateway)
-		if ipv6gw.Equal(gwIP.IPv6Address.IP()) {
-			ip, err := addressing.NewCiliumIPv6(contNetwork.GlobalIPv6Address)
-			if err == nil {
-				return ip.EndpointID()
-			}
-		}
+	if cont.NetworkSettings == nil {
+		return 0
 	}
 
+	if ciliumIP := getCiliumIPv6(cont.NetworkSettings.Networks, gwIP); ciliumIP != nil {
+		return ciliumIP.EndpointID()
+	}
 	return 0
 }
 
+func getCiliumIPv6(networks map[string]*dNetwork.EndpointSettings, gwIP *addressing.NodeAddress) *addressing.CiliumIPv6 {
```

Comment:
I check this, the map have the name (not type) as key and the endpoint settings as values. The user can decide whatever network we would like to the network name, for example:
`docker network create --ipv6 --subnet ::1/112 --ipam-driver cilium --driver cilium foo-cilium`
While inspecting the container you will see something like:
```
            "Networks": {
                "foo-cilium": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "42062ac9e052"
                    ],
                    "NetworkID": "006d05ef51786264e63bc3cdaea378ceb49b4bf0e5518c25b49f86c5ec3d474a",
                    "EndpointID": "816c59aa0408ff543fa20fde30ec1b038ebc3c958acb10690f97308af75e12ef",
                    "Gateway": "",
                    "IPAddress": "10.15.116.202",
                    "IPPrefixLen": 32,
                    "IPv6Gateway": "f00d::a00:20f:0:0",
                    "GlobalIPv6Address": "f00d::a00:20f:0:f236",
                    "GlobalIPv6PrefixLen": 128,
                    "MacAddress": ""
                }
            }
```
Regarding the conflicting IPs across docker networks, I was able to create conflicting networks but I think it was because we are assigning the subnet `0.0.0.0/0` to cilium

```vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ docker network create --subnet=10.15.0.0/16 --ip-range 10.15.0.0/24 --gateway 10.15.0.1 --driver bridge  conflicting-ips  
46926b026d890a1a6589b4efdf42a2dda5a13accd611b469cc2b0844ca6ad91d
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
0cb2b91ae70c        bridge              bridge              local               
46926b026d89        conflicting-ips     bridge              local               
006d05ef5178        foo-cilium          cilium              local               
1fc1aa3113cc        foo-cilium-2        cilium              local               
99abdbf2eba0        host                host                local               
6e7c120bf9bf        none                null                local               
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ docker network inspect conflicting-ips
[
    {
        "Name": "conflicting-ips",
        "Id": "46926b026d890a1a6589b4efdf42a2dda5a13accd611b469cc2b0844ca6ad91d",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": {},
            "Config": [
                {
                    "Subnet": "10.15.0.0/16",
                    "IPRange": "10.15.0.0/24",
                    "Gateway": "10.15.0.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {},
        "Options": {},
        "Labels": {}
    }
]
$ docker network inspect foo-cilium-2
[
    {
        "Name": "foo-cilium-2",
        "Id": "1fc1aa3113cc19c67fa030f6886c4dd7afb1b14778489644fba3b53b1925684a",
        "Scope": "local",
        "Driver": "cilium",
        "EnableIPv6": true,
        "IPAM": {
            "Driver": "cilium",
            "Options": {},
            "Config": [
                {
                    "Subnet": "0.0.0.0/0",
                    "Gateway": "10.15.0.1/32"
                },
                {
                    "Subnet": "::1/112"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "a5f8cdb23f9e41064f30b7e2aef18e736625a00066c96145e50f34b3fe85dc38": {
                "Name": "server-4",
                "EndpointID": "403e09230878b4c0b0584da0a5f2b0287cbab038504ef618835c13daf1e6b4d2",
                "MacAddress": "",
                "IPv4Address": "10.15.129.91/32",
                "IPv6Address": "f00d::a00:20f:0:dc06/128"
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```